New implementation of #633

bricerebsamen opened the pull request:
This is a new implementation of #633. I did it on top of recent changes in indigo-devel, with the changes from the catkin_download branch. I also modified @trainman419 's implementation a bit to take into account @dirk-thomas requests.

Hopefully this will be finally merged and will fix #690.

See discussion for #633 and #690 for info

dirk-thomas committed:
separate download function from tests (#633)

bricerebsamen committed:
Merge remote-tracking branch 'origin/catkin_download' into download_data_new

Bringing old branch catkin_download up to date

bricerebsamen committed:
fixed typo in cmake parameter (url)

bricerebsamen committed:
Add catkin_download_(build|extra)_data functions

Sort of re-implemented what @trainman419 did in trainman419/catkin@ab93b61
(see PR #633) on top of more recent changes. This should allow to merge more
easily.

On top of bringing more flexibility for downloading data, it also fixes #690

I made a few modifications though:

- In @trainman419's implementation, the destination would be set to the
  devel space, whereas the original implentation used the build space.
  I kept the original design, because otherwise it builds a bunch of tests.

- as requested by @dirk-thomas, the target is not an optional argument.

dirk-thomas commented on cmake/catkin_download.cmake:
```diff
@@ -0,0 +1,146 @@
+# :type url: string
+#
+# :param DESTINATION: the directory where the file is downloaded to
+#   (default: ${CATKIN_DEVEL_PREFIX}/${CATKIN_PACKAGE_SHARE_DESTINATION})
```
The default value in the code seems to be different: https://github.com/ros/catkin/pull/716/files#diff-f6d46f8e02eefd6f88ad660967bc8182R44

dirk-thomas commented on cmake/catkin_download.cmake:
```diff
@@ -0,0 +1,146 @@
+# :type MD5: string
+# :param EXCLUDE_FROM_ALL: exclude this download from the 'all' build target
+#   (default: true, excluding it from the all target)
+# :type EXCLUDE_FROM_ALL: bool
```
In CMake `EXCLUDE_FROM_ALL` is usually used as an option. It might be confusing to use it as a boolean argument.

dirk-thomas commented on cmake/catkin_download.cmake:
```diff
@@ -0,0 +1,146 @@
+# :type EXCLUDE_FROM_ALL: bool
+# :param REQUIRED: require the file to be available for the build to succeed
+#   (default: true, requiring the file)
+# :type REQUIRED: bool
```
Same as `EXCLUDE_FROM_ALL`. CMake uses `REQUIRED` usually as an option.

bricerebsamen replied:
can someone provide some guidance or examples on how to do this properly. I spent the last 30 minutes searching but found nothing.

dirk-thomas commented on cmake/catkin_download.cmake:
```diff
@@ -0,0 +1,146 @@
+# @public
+function(catkin_download_extra_data)
+  catkin_download(${ARGN} EXCLUDE_FROM_ALL true)
+endfunction()
```
Do we really need these two functions (`catkin_download_build_data` and `catkin_download_extra_data`) since they only pass a single argument to the download function? I would rather avoid adding those in favor of documenting the uses when how to invoke `catkin_download` for different use cases.

dirk-thomas commented on cmake/catkin_download.cmake:
```diff
@@ -0,0 +1,146 @@
+endfunction()
+
+if(NOT TARGET download_extra_data)
+  add_custom_target(download_extra_data)
```
What is this target being used for?

bricerebsamen replied:
it's referenced at line 70. All the downloads will be added to this target. So that if you want to download the data you can call this target. This is useful if the download targets were added with EXCLUDE_FROM_ALL. This way you can build all first, and the call download_extra_data instead of calling all the download targets one by one.

bricerebsamen committed:
some improvements to the download functions

- addressed @dirk-thomas comments in #716:
  - corrected the doc for the default destination to match the code
  - REQUIRED and EXCLUDE_FROM_ALL as options instead of arguments
  - removed the extra functions catkin_download_build_data and
    catkin_download_extra_data
- fixed a bug with the --ignore-error option to download_checkmd5.py
- in catkin_download_test_data(), calling catkin_download() with the
  EXCLUDE_FROM_ALL option.

bricerebsamen commented:
is there something stopping this PR from being merged? anything I can do?

wjwwood commented:
We probably didn't notice it had been updated. We only get emails when you comment, not when you push a commit. If that's not the case, then we just haven't had time to give it a second review since the most recent changes.

I'll give it a review now.

wjwwood commented:
I reviewed the changes in ec239b6, but not the whole request, and they look reasonable and in line with @dirk-thomas's comments.

dirk-thomas commented:
Code review looks good. I will try to run and merge it sometime next week.

trainman419 commented:
Bump. Has anyone had a chance to test and validate this?

dirk-thomas commented:
Not yet, it is scheduled for the "ROS bug / ticket squashing day" next Thursday.

dirk-thomas commented:
Thank you for the PR!

Squashed and merged with only minor changes: 09ce538e47e179dbd433d9c6a2f04211c609a44b
